src/Latte/Engine.php

```diff
@@ -156,33 +156,53 @@ private function loadCacheFile($name, $params)
 		}
 
 		$file = $this->getCacheFile($name);
-		$handle = fopen($file, 'c+');
+		if ($this->autoRefresh && $this->getLoader()->isExpired($name, (int) @filemtime($file))) {
+			@unlink($file);
+
+		} elseif (self::tryInclude($file, $params)) {
+			return;
+		}
+
+		if (!is_dir($this->tempDirectory)) {
+			@mkdir($this->tempDirectory);
+		}
+
+		$handle = fopen("$file.lock", 'c+');
 		if (!$handle) {
-			throw new \RuntimeException("Unable to open or create file '$file'.");
+			throw new \RuntimeException("Unable to open or create file '$file.lock'.");
 		}
-		flock($handle, LOCK_SH);
-		$stat = fstat($handle);
-		if (!$stat['size'] || ($this->autoRefresh && $this->getLoader()->isExpired($name, $stat['mtime']))) {
-			ftruncate($handle, 0);
-			flock($handle, LOCK_EX);
-			$stat = fstat($handle);
-			if (!$stat['size']) {
-				$code = $this->compile($name);
-				if (fwrite($handle, $code, strlen($code)) !== strlen($code)) {
-					ftruncate($handle, 0);
-					throw new \RuntimeException("Unable to write file '$file'.");
-				}
+
+		if (flock($handle, LOCK_EX) && !is_file($file)) {
+			$code = $this->compile($name);
```

Comment:
Which may in theory damage the compiled file because each of those thread may write something different if source template will change in between compile calls. I think that exception is fine because `flock` should never time out it.